hivemq-edge/src/main/java/com/hivemq/HiveMQEdgeBootstrap.java

```diff
@@ -0,0 +1,229 @@
+    private final @NotNull RestComponentsService restComponentsService =
+            new RestComponentsServiceImpl(genericAPIHolder);
+    private @Nullable CommercialModuleLoaderDiscovery commercialModuleLoaderDiscovery;
+    private @Nullable GeneralBootstrapServiceImpl generalBootstrapService;
+    private @Nullable PersistenceBootstrapService persistenceBootstrapService;
+    private @Nullable Injector injector;
+
+    public HiveMQEdgeBootstrap(
+            final @NotNull MetricRegistry metricRegistry,
+            final @NotNull SystemInformation systemInformation,
+            final @NotNull ModuleLoader moduleLoader,
+            final @Nullable ConfigurationService configService) {
+        this.metricRegistry = metricRegistry;
+        this.systemInformation = systemInformation;
+        this.moduleLoader = moduleLoader;
+        this.configService = configService;
+    }
+
+    public @NotNull Injector bootstrap() throws HiveMQEdgeStartupException {
+        metricRegistry.addListener(new MetricRegistryLogger());
+
+        LoggingBootstrap.prepareLogging();
+
+        // Embedded has already called init as it is required to read the config file.
+        if (!systemInformation.isEmbedded()) {
+            log.trace("Initializing HiveMQ home directory");
+            //Create SystemInformation this early because logging depends on it
+            systemInformation.init();
+        }
+
+        log.trace("Initializing Logging");
+        LoggingBootstrap.initLogging(systemInformation.getConfigFolder());
+
+        log.trace("Initializing Exception handlers");
+        HiveMQExceptionHandlerBootstrap.addUnrecoverableExceptionHandler();
+
+        //ungraceful shutdown does not delete tmp folders, so we clean them up on broker start
+        log.trace("Cleaning up temporary folders");
+        deleteTmpFolder(systemInformation.getDataFolder());
+
+        if (configService == null) {
+            log.trace("Initializing configuration");
+            configService = ConfigurationBootstrap.bootstrapConfig(systemInformation);
+        }
+        bootstrapCoreComponents();
+
+        bootstrapInjector();
+        final long startInit = System.currentTimeMillis();
+
+        bootstrapPersistences();
+
+        // make sure all persistences are bootstrapped.
+        awaitPersistenceStartup();
+
+        finishBootstrap();
+
+        log.trace("Initializing classes");
+        Objects.requireNonNull(injector).initEagerSingletons();
+        log.trace("Initialized classes in {}ms", (System.currentTimeMillis() - startInit));
+        return injector;
+    }
+
+    private void awaitPersistenceStartup() {
+        Objects.requireNonNull(injector).persistences();
+        try {
+            persistenceStartup.finish();
+        } catch (InterruptedException e) {
+            throw new HiveMQEdgeStartupException(e);
+        }
+
+        // configService is always set in caller
+        assert configService != null;
```

Comment:
Why isn't this asserted at the begin of the method?

Reply:
yeah, good point. I streamline it with Preconditions